File: backend/app/api/v1/endpoints/knowledge_graph.py
import uuid
from typing import Optional, List
from fastapi import APIRouter, Depends, Query, status
from sqlalchemy.orm import Session
from loguru import logger

from app.database.session import get_db
from app.api.dependencies import get_current_user, check_rate_limit
from app.models.user import User
from app.api.v1.endpoints.documents import resolve_workspace_id
from app.services.knowledge_graph import KnowledgeGraphService, NodeNotFound, EdgeNotFound
from app.services.knowledge_graph_intelligence import KnowledgeGraphIntelligenceService
from app.schemas.knowledge_graph import (
    NodeCreate,
    NodeUpdate,
    NodeResponse,
    EdgeCreate,
    EdgeResponse,
    NeighborResponse,
    GraphTraversalRequest,
    GraphTraversalResponse,
    GraphContextResponse,
    RelatedEntitiesResponse,
    PathSearchRequest,
    GraphPathResponse,
    RelationshipAnalysisRequest,
    RelationshipAnalysisResponse,
    GraphIntelligenceContextRequest,
    RelatedEntityItem
)

router = APIRouter(prefix="/knowledge-graph", tags=["Knowledge Graph"])

# ---------------------------------------------------------
# Node Endpoints
# ---------------------------------------------------------

@router.post("/nodes", response_model=NodeResponse, status_code=status.HTTP_201_CREATED, dependencies=[Depends(check_rate_limit)])
def create_node_endpoint(
    payload: NodeCreate,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    workspace_id = resolve_workspace_id(current_user, db)
    service = KnowledgeGraphService(db)
    return service.create_node(
        user_id=current_user.id,
        workspace_id=workspace_id,
        node_data=payload
    )

@router.get("/nodes", response_model=List[NodeResponse], dependencies=[Depends(check_rate_limit)])
def list_nodes_endpoint(
    node_type: Optional[str] = Query(None),
    external_id: Optional[str] = Query(None),
    skip: int = Query(0, ge=0),
    limit: int = Query(50, ge=1, le=100),
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    workspace_id = resolve_workspace_id(current_user, db)
    service = KnowledgeGraphService(db)
    nodes, _ = service.list_nodes(
        user_id=current_user.id,
        workspace_id=workspace_id,
        node_type=node_type,
        external_id=external_id,
        skip=skip,
        limit=limit
    )
    return nodes

@router.get("/nodes/{node_id}", response_model=NodeResponse, dependencies=[Depends(check_rate_limit)])
def get_node_endpoint(
    node_id: uuid.UUID,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    workspace_id = resolve_workspace_id(current_user, db)
    service = KnowledgeGraphService(db)
    node = service.get_node(user_id=current_user.id, workspace_id=workspace_id, node_id=node_id)
    if not node:
        raise NodeNotFound(f"Node {node_id} not found in workspace.")
    return node

@router.patch("/nodes/{node_id}", response_model=NodeResponse, dependencies=[Depends(check_rate_limit)])
def update_node_endpoint(
    node_id: uuid.UUID,
    payload: NodeUpdate,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    workspace_id = resolve_workspace_id(current_user, db)
    service = KnowledgeGraphService(db)
    return service.update_node(
        user_id=current_user.id,
        workspace_id=workspace_id,
        node_id=node_id,
        update_data=payload
    )

@router.delete("/nodes/{node_id}", status_code=status.HTTP_204_NO_CONTENT, dependencies=[Depends(check_rate_limit)])
def delete_node_endpoint(
    node_id: uuid.UUID,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    workspace_id = resolve_workspace_id(current_user, db)
    service = KnowledgeGraphService(db)
    service.delete_node(user_id=current_user.id, workspace_id=workspace_id, node_id=node_id)
    return None

# ---------------------------------------------------------
# Edge Endpoints
# ---------------------------------------------------------

@router.post("/edges", response_model=EdgeResponse, status_code=status.HTTP_201_CREATED, dependencies=[Depends(check_rate_limit)])
def create_edge_endpoint(
    payload: EdgeCreate,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    workspace_id = resolve_workspace_id(current_user, db)
    service = KnowledgeGraphService(db)
    return service.create_edge(
        user_id=current_user.id,
        workspace_id=workspace_id,
        edge_data=payload
    )

@router.get("/edges", response_model=List[EdgeResponse], dependencies=[Depends(check_rate_limit)])
def list_edges_endpoint(
    relationship_type: Optional[str] = Query(None),
    source_node_id: Optional[uuid.UUID] = Query(None),
    target_node_id: Optional[uuid.UUID] = Query(None),
    skip: int = Query(0, ge=0),
    limit: int = Query(50, ge=1, le=100),
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    workspace_id = resolve_workspace_id(current_user, db)
    service = KnowledgeGraphService(db)
    edges, _ = service.list_edges(
        user_id=current_user.id,
        workspace_id=workspace_id,
        relationship_type=relationship_type,
        source_node_id=source_node_id,
        target_node_id=target_node_id,
        skip=skip,
        limit=limit
    )
    return edges

@router.get("/edges/{edge_id}", response_model=EdgeResponse, dependencies=[Depends(check_rate_limit)])
def get_edge_endpoint(
    edge_id: uuid.UUID,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    workspace_id = resolve_workspace_id(current_user, db)
    service = KnowledgeGraphService(db)
    edge = service.get_edge(user_id=current_user.id, workspace_id=workspace_id, edge_id=edge_id)
    if not edge:
        raise EdgeNotFound(f"Edge {edge_id} not found in workspace.")
    return edge

@router.delete("/edges/{edge_id}", status_code=status.HTTP_204_NO_CONTENT, dependencies=[Depends(check_rate_limit)])
def delete_edge_endpoint(
    edge_id: uuid.UUID,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    workspace_id = resolve_workspace_id(current_user, db)
    service = KnowledgeGraphService(db)
    service.delete_edge(user_id=current_user.id, workspace_id=workspace_id, edge_id=edge_id)
    return None

# ---------------------------------------------------------
# Graph Traversal, Neighbors & Basic Search
# ---------------------------------------------------------

@router.get("/nodes/{node_id}/neighbors", response_model=List[NeighborResponse], dependencies=[Depends(check_rate_limit)])
def get_neighbors_endpoint(
    node_id: uuid.UUID,
    relationship_types: Optional[List[str]] = Query(None),
    direction: str = Query("both", pattern="^(both|outgoing|incoming)$"),
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    workspace_id = resolve_workspace_id(current_user, db)
    service = KnowledgeGraphService(db)
    return service.get_neighbors(
        user_id=current_user.id,
        workspace_id=workspace_id,
        node_id=node_id,
        relationship_types=relationship_types,
        direction=direction
    )

@router.post("/traverse", response_model=GraphTraversalResponse, dependencies=[Depends(check_rate_limit)])
def traverse_endpoint(
    payload: GraphTraversalRequest,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    workspace_id = resolve_workspace_id(current_user, db)
    service = KnowledgeGraphService(db)
    return service.traverse(
        user_id=current_user.id,
        workspace_id=workspace_id,
        start_node_ids=payload.start_node_ids,
        max_depth=payload.max_depth,
        relationship_types=payload.relationship_types,
        node_types=payload.node_types,
        limit=payload.limit
    )

@router.get("/search", response_model=List[NodeResponse], dependencies=[Depends(check_rate_limit)])
def search_nodes_endpoint(
    q: str = Query(..., min_length=1),
    node_type: Optional[str] = Query(None),
    limit: int = Query(20, ge=1, le=100),
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    workspace_id = resolve_workspace_id(current_user, db)
    service = KnowledgeGraphService(db)
    return service.search_nodes(
        user_id=current_user.id,
        workspace_id=workspace_id,
        query=q,
        node_type=node_type,
        limit=limit
    )

# ---------------------------------------------------------
# Phase 5.1: Graph Intelligence Endpoints
# ---------------------------------------------------------

@router.get("/nodes/{node_id}/related", response_model=RelatedEntitiesResponse, dependencies=[Depends(check_rate_limit)])
def get_related_entities_endpoint(
    node_id: uuid.UUID,
    depth: int = Query(2, ge=1, le=5),
    limit: int = Query(50, ge=1, le=500),
    relationship_types: Optional[List[str]] = Query(None),
    node_types: Optional[List[str]] = Query(None),
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    workspace_id = resolve_workspace_id(current_user, db)
    intelligence_service = KnowledgeGraphIntelligenceService(db)
    return intelligence_service.get_related_entities(
        user_id=current_user.id,
        workspace_id=workspace_id,
        node_id=node_id,
        depth=depth,
        limit=limit,
        relationship_types=relationship_types,
        node_types=node_types
    )

@router.post("/path", response_model=GraphPathResponse, dependencies=[Depends(check_rate_limit)])
def find_shortest_path_endpoint(
    payload: PathSearchRequest,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    workspace_id = resolve_workspace_id(current_user, db)
    intelligence_service = KnowledgeGraphIntelligenceService(db)
    return intelligence_service.find_shortest_path(
        user_id=current_user.id,
        workspace_id=workspace_id,
        source_node_id=payload.source_node_id,
        target_node_id=payload.target_node_id,
        max_depth=payload.max_depth,
        allowed_relationship_types=payload.allowed_relationship_types
    )

@router.post("/analyze", response_model=RelationshipAnalysisResponse, dependencies=[Depends(check_rate_limit)])
def analyze_relationships_endpoint(
    payload: RelationshipAnalysisRequest,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    workspace_id = resolve_workspace_id(current_user, db)
    intelligence_service = KnowledgeGraphIntelligenceService(db)
    return intelligence_service.analyze_relationships(
        user_id=current_user.id,
        workspace_id=workspace_id,
        source_node_id=payload.source_node_id,
        target_node_id=payload.target_node_id,
        max_depth=payload.max_depth
    )

@router.post("/context", response_model=GraphContextResponse, dependencies=[Depends(check_rate_limit)])
def get_graph_context_endpoint(
    payload: GraphIntelligenceContextRequest,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    workspace_id = resolve_workspace_id(current_user, db)
    intelligence_service = KnowledgeGraphIntelligenceService(db)
    formatted = intelligence_service.build_graph_context(
        user_id=current_user.id,
        workspace_id=workspace_id,
        entity_names=payload.entity_names,
        node_ids=payload.node_ids,
        depth=payload.depth,
        max_entities=payload.max_entities
    )
    return GraphContextResponse(
        entities=[],
        relationships=[],
        formatted_context=formatted
    )

@router.get("/search/enhanced", response_model=List[RelatedEntityItem], dependencies=[Depends(check_rate_limit)])
def enhanced_graph_search_endpoint(
    q: Optional[str] = Query(None),
    node_type: Optional[str] = Query(None),
    relationship_type: Optional[str] = Query(None),
    depth: int = Query(1, ge=0, le=3),
    limit: int = Query(50, ge=1, le=100),
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    workspace_id = resolve_workspace_id(current_user, db)
    intelligence_service = KnowledgeGraphIntelligenceService(db)
    return intelligence_service.enhanced_graph_search(
        user_id=current_user.id,
        workspace_id=workspace_id,
        query=q,
        node_type=node_type,
        relationship_type=relationship_type,
        depth=depth,
        limit=limit
    )

# ---------------------------------------------------------
# Phase 5.2: Document Entity Extraction & Graph Construction Endpoints
# ---------------------------------------------------------

@router.post("/documents/{document_id}/extract", response_model=dict, dependencies=[Depends(check_rate_limit)])
def extract_document_graph_endpoint(
    document_id: uuid.UUID,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """
    Extracts entities and constructs knowledge graph representation for the document.
    """
    from app.services.graph_construction import GraphConstructionService
    workspace_id = resolve_workspace_id(current_user, db)
    construction_service = GraphConstructionService(db)
    return construction_service.construct_graph_from_document(
        document_id=document_id,
        user_id=current_user.id,
        workspace_id=workspace_id
    )

@router.get("/documents/{document_id}/entities", response_model=List[NodeResponse], dependencies=[Depends(check_rate_limit)])
def get_document_entities_endpoint(
    document_id: uuid.UUID,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """
    Returns all knowledge graph entities associated with the document.
    """
    from app.services.graph_construction import GraphConstructionService
    workspace_id = resolve_workspace_id(current_user, db)
    construction_service = GraphConstructionService(db)
    return construction_service.get_document_entities(
        document_id=document_id,
        user_id=current_user.id,
        workspace_id=workspace_id
    )

@router.get("/documents/{document_id}/relationships", response_model=List[EdgeResponse], dependencies=[Depends(check_rate_limit)])
def get_document_relationships_endpoint(
    document_id: uuid.UUID,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """
    Returns all knowledge graph edge relationships for the document.
    """
    from app.services.graph_construction import GraphConstructionService
    workspace_id = resolve_workspace_id(current_user, db)
    construction_service = GraphConstructionService(db)
    return construction_service.get_document_relationships(
        document_id=document_id,
        user_id=current_user.id,
        workspace_id=workspace_id
    )

@router.post("/documents/{document_id}/rebuild", response_model=dict, dependencies=[Depends(check_rate_limit)])
def rebuild_document_graph_endpoint(
    document_id: uuid.UUID,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """
    Safely rebuilds the knowledge graph nodes and edges for the specified document.
    """
    from app.services.graph_construction import GraphConstructionService
    workspace_id = resolve_workspace_id(current_user, db)
    construction_service = GraphConstructionService(db)
    return construction_service.rebuild_document_graph(
        document_id=document_id,
        user_id=current_user.id,
        workspace_id=workspace_id
    )

# ---------------------------------------------------------
# Memory-Graph Synchronization Endpoints
# ---------------------------------------------------------

@router.post("/sync/memory/{memory_id}", response_model=dict, dependencies=[Depends(check_rate_limit)])
def sync_memory_to_graph_endpoint(
    memory_id: uuid.UUID,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """
    Synchronizes an AgentMemory record into the tenant Knowledge Graph.
    """
    from app.models.memory import AgentMemory
    from app.services.memory_graph_sync import MemoryGraphSyncService
    workspace_id = resolve_workspace_id(current_user, db)

    memory = db.query(AgentMemory).filter(
        AgentMemory.id == memory_id,
        AgentMemory.workspace_id == workspace_id,
        AgentMemory.user_id == current_user.id
    ).first()

    if not memory:
        from fastapi import HTTPException
        raise HTTPException(status_code=404, detail="Memory record not found in workspace.")

    sync_service = MemoryGraphSyncService(db)
    return sync_service.sync_memory_to_graph(
        user_id=current_user.id,
        workspace_id=workspace_id,
        memory=memory
    )

@router.post("/nodes/{node_id}/sync-memory", response_model=dict, dependencies=[Depends(check_rate_limit)])
def sync_graph_node_to_memory_endpoint(
    node_id: uuid.UUID,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """
    Synchronizes a KnowledgeGraphNode entity into an AgentMemory record.
    """
    from app.services.memory_graph_sync import MemoryGraphSyncService
    workspace_id = resolve_workspace_id(current_user, db)

    sync_service = MemoryGraphSyncService(db)
    mem = sync_service.sync_graph_to_memory(
        user_id=current_user.id,
        workspace_id=workspace_id,
        node_id=node_id
    )
    if not mem:
        from fastapi import HTTPException
        raise HTTPException(status_code=404, detail="Node not found or sync skipped to prevent loop.")

    return {
        "status": "synced",
        "memory_id": str(mem.id),
        "content": mem.content,
        "source": mem.source
    }

# ---------------------------------------------------------
# Phase 5.8: Graph Search & Analytics Endpoints
# ---------------------------------------------------------

@router.get("/analytics/overview", response_model=dict, dependencies=[Depends(check_rate_limit)])
def get_graph_analytics_overview_endpoint(
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """
    Returns high-level graph topology, degree distribution, and provenance analytics.
    """
    from app.services.graph_analytics import GraphAnalyticsService
    workspace_id = resolve_workspace_id(current_user, db)
    analytics = GraphAnalyticsService(db)
    return analytics.get_analytics_overview(user_id=current_user.id, workspace_id=workspace_id).model_dump()

@router.get("/analytics/health", response_model=dict, dependencies=[Depends(check_rate_limit)])
def get_graph_health_endpoint(
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """
    Returns graph health diagnostics, orphan rates, and contradiction indicators.
    """
    from app.services.graph_analytics import GraphAnalyticsService
    workspace_id = resolve_workspace_id(current_user, db)
    analytics = GraphAnalyticsService(db)
    return analytics.get_graph_health(user_id=current_user.id, workspace_id=workspace_id).model_dump()

@router.get("/analytics/top-connected", response_model=List[dict], dependencies=[Depends(check_rate_limit)])
def get_top_connected_entities_endpoint(
    limit: int = Query(10, ge=1, le=50),
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """
    Returns the most connected hub nodes in the workspace graph.
    """
    from app.services.graph_analytics import GraphAnalyticsService
    workspace_id = resolve_workspace_id(current_user, db)
    analytics = GraphAnalyticsService(db)
    items = analytics.get_top_connected_entities(user_id=current_user.id, workspace_id=workspace_id, limit=limit)
    return [i.model_dump() for i in items]

@router.get("/analytics/orphans", response_model=List[dict], dependencies=[Depends(check_rate_limit)])
def get_orphan_nodes_endpoint(
    limit: int = Query(50, ge=1, le=100),
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """
    Returns isolated entities with no connected relationships.
    """
    from app.services.graph_analytics import GraphAnalyticsService
    workspace_id = resolve_workspace_id(current_user, db)
    analytics = GraphAnalyticsService(db)
    orphans = analytics.get_orphan_nodes(user_id=current_user.id, workspace_id=workspace_id, limit=limit)
    return [o.model_dump() for o in orphans]

@router.get("/analytics/duplicates", response_model=List[dict], dependencies=[Depends(check_rate_limit)])
def get_duplicate_candidates_endpoint(
    similarity_threshold: float = Query(0.85, ge=0.5, le=1.0),
    limit: int = Query(20, ge=1, le=50),
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """
    Returns potential duplicate entity pairs for safe review and merging.
    """
    from app.services.graph_analytics import GraphAnalyticsService
    workspace_id = resolve_workspace_id(current_user, db)
    analytics = GraphAnalyticsService(db)
    candidates = analytics.detect_duplicate_candidates(
        user_id=current_user.id,
        workspace_id=workspace_id,
        similarity_threshold=similarity_threshold,
        limit=limit
    )
    return [c.model_dump() for c in candidates]

@router.post("/search/advanced", response_model=dict, dependencies=[Depends(check_rate_limit)])
def advanced_search_endpoint(
    payload: dict,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """
    Executes advanced multi-criterion search with deterministic relevance ranking.
    """
    from app.schemas.knowledge_graph import AdvancedSearchRequest
    from app.services.graph_analytics import GraphAnalyticsService
    workspace_id = resolve_workspace_id(current_user, db)
    req = AdvancedSearchRequest.model_validate(payload)
    analytics = GraphAnalyticsService(db)
    return analytics.advanced_search(user_id=current_user.id, workspace_id=workspace_id, req=req).model_dump()

File: frontend/src/api/knowledgeGraph.ts
import { request } from './client';

export interface KnowledgeGraphNode {
  id: string;
  user_id: string;
  workspace_id: string;
  node_type: string;
  external_id?: string | null;
  name: string;
  description?: string | null;
  metadata?: Record<string, any> | null;
  created_at: string;
  updated_at: string;
}

export interface KnowledgeGraphEdge {
  id: string;
  user_id: string;
  workspace_id: string;
  source_node_id: string;
  target_node_id: string;
  relationship_type: string;
  confidence: number;
  properties?: Record<string, any> | null;
  created_at: string;
  updated_at: string;
}

export interface NeighborItem {
  node: KnowledgeGraphNode;
  relationship_type: string;
  direction: 'outgoing' | 'incoming';
  confidence: number;
  edge_id: string;
}

export interface RelatedEntityItem {
  node_id: string;
  node_type: string;
  name: string;
  description?: string | null;
  distance: number;
  relevance_score: number;
  relationship_path: string[];
  metadata?: Record<string, any> | null;
}

export interface RelatedEntitiesResponse {
  node_id: string;
  depth: number;
  total_related: number;
  related_entities: RelatedEntityItem[];
}

export interface GraphPathStep {
  from_node_id: string;
  from_node_name: string;
  to_node_id: string;
  to_node_name: string;
  relationship_type: string;
  direction: 'outgoing' | 'incoming';
  confidence: number;
}

export interface GraphPathResponse {
  source_node_id: string;
  target_node_id: string;
  path_found: boolean;
  distance: number;
  steps: GraphPathStep[];
  nodes: KnowledgeGraphNode[];
}

export interface RelationshipDetail {
  relationship_type: string;
  direction: 'outgoing' | 'incoming' | 'bidirectional';
  confidence: number;
  distance: number;
  via_nodes: string[];
}

export interface RelationshipAnalysisResponse {
  source_node: KnowledgeGraphNode;
  target_node: KnowledgeGraphNode;
  are_connected: boolean;
  min_distance?: number | null;
  direct_relationships: RelationshipDetail[];
  indirect_relationships: RelationshipDetail[];
  summary: string;
}

export interface GraphContextResponse {
  entities: any[];
  relationships: any[];
  formatted_context: string;
}

export interface GraphTraversalResponse {
  nodes: KnowledgeGraphNode[];
  edges: KnowledgeGraphEdge[];
  depth_reached: number;
  total_nodes: number;
  total_edges: number;
}

// ----------------------------------------------------------------------
// Core API Calls
// ----------------------------------------------------------------------

export async function listNodes(params?: {
  node_type?: string;
  external_id?: string;
  skip?: number;
  limit?: number;
}): Promise<KnowledgeGraphNode[]> {
  const query = new URLSearchParams();
  if (params?.node_type) query.append('node_type', params.node_type);
  if (params?.external_id) query.append('external_id', params.external_id);
  if (params?.skip !== undefined) query.append('skip', String(params.skip));
  if (params?.limit !== undefined) query.append('limit', String(params.limit));

  const qs = query.toString();
  return request<KnowledgeGraphNode[]>(`/knowledge-graph/nodes${qs ? `?${qs}` : ''}`, {
    method: 'GET',
  });
}

export async function getNode(nodeId: string): Promise<KnowledgeGraphNode> {
  return request<KnowledgeGraphNode>(`/knowledge-graph/nodes/${nodeId}`, {
    method: 'GET',
  });
}

export async function createNode(payload: {
  node_type: string;
  name: string;
  external_id?: string;
  description?: string;
  metadata?: Record<string, any>;
}): Promise<KnowledgeGraphNode> {
  return request<KnowledgeGraphNode>('/knowledge-graph/nodes', {
    method: 'POST',
    body: JSON.stringify(payload),
  });
}

export async function deleteNode(nodeId: string): Promise<void> {
  return request<void>(`/knowledge-graph/nodes/${nodeId}`, {
    method: 'DELETE',
  });
}

export async function listEdges(params?: {
  relationship_type?: string;
  source_node_id?: string;
  target_node_id?: string;
  skip?: number;
  limit?: number;
}): Promise<KnowledgeGraphEdge[]> {
  const query = new URLSearchParams();
  if (params?.relationship_type) query.append('relationship_type', params.relationship_type);
  if (params?.source_node_id) query.append('source_node_id', params.source_node_id);
  if (params?.target_node_id) query.append('target_node_id', params.target_node_id);
  if (params?.skip !== undefined) query.append('skip', String(params.skip));
  if (params?.limit !== undefined) query.append('limit', String(params.limit));

  const qs = query.toString();
  return request<KnowledgeGraphEdge[]>(`/knowledge-graph/edges${qs ? `?${qs}` : ''}`, {
    method: 'GET',
  });
}

export async function createEdge(payload: {
  source_node_id: string;
  target_node_id: string;
  relationship_type: string;
  confidence?: number;
  properties?: Record<string, any>;
}): Promise<KnowledgeGraphEdge> {
  return request<KnowledgeGraphEdge>('/knowledge-graph/edges', {
    method: 'POST',
    body: JSON.stringify(payload),
  });
}

export async function deleteEdge(edgeId: string): Promise<void> {
  return request<void>(`/knowledge-graph/edges/${edgeId}`, {
    method: 'DELETE',
  });
}

export async function getNeighbors(
  nodeId: string,
  params?: { relationship_types?: string[]; direction?: string }
): Promise<NeighborItem[]> {
  const query = new URLSearchParams();
  if (params?.direction) query.append('direction', params.direction);
  if (params?.relationship_types) {
    params.relationship_types.forEach((r) => query.append('relationship_types', r));
  }

  const qs = query.toString();
  return request<NeighborItem[]>(`/knowledge-graph/nodes/${nodeId}/neighbors${qs ? `?${qs}` : ''}`, {
    method: 'GET',
  });
}

export async function getRelatedEntities(
  nodeId: string,
  params?: {
    depth?: number;
    limit?: number;
    relationship_types?: string[];
    node_types?: string[];
  }
): Promise<RelatedEntitiesResponse> {
  const query = new URLSearchParams();
  if (params?.depth !== undefined) query.append('depth', String(params.depth));
  if (params?.limit !== undefined) query.append('limit', String(params.limit));
  if (params?.relationship_types) {
    params.relationship_types.forEach((r) => query.append('relationship_types', r));
  }
  if (params?.node_types) {
    params.node_types.forEach((nt) => query.append('node_types', nt));
  }

  const qs = query.toString();
  return request<RelatedEntitiesResponse>(`/knowledge-graph/nodes/${nodeId}/related${qs ? `?${qs}` : ''}`, {
    method: 'GET',
  });
}

export async function searchNodes(query: string, nodeType?: string, limit: number = 20): Promise<KnowledgeGraphNode[]> {
  const q = new URLSearchParams({ q: query, limit: String(limit) });
  if (nodeType) q.append('node_type', nodeType);
  return request<KnowledgeGraphNode[]>(`/knowledge-graph/search?${q.toString()}`, {
    method: 'GET',
  });
}

export async function searchEnhanced(params: {
  q?: string;
  node_type?: string;
  relationship_type?: string;
  depth?: number;
  limit?: number;
}): Promise<RelatedEntityItem[]> {
  const query = new URLSearchParams();
  if (params.q) query.append('q', params.q);
  if (params.node_type) query.append('node_type', params.node_type);
  if (params.relationship_type) query.append('relationship_type', params.relationship_type);
  if (params.depth !== undefined) query.append('depth', String(params.depth));
  if (params.limit !== undefined) query.append('limit', String(params.limit));

  const qs = query.toString();
  return request<RelatedEntityItem[]>(`/knowledge-graph/search/enhanced${qs ? `?${qs}` : ''}`, {
    method: 'GET',
  });
}

export async function findPath(payload: {
  source_node_id: string;
  target_node_id: string;
  max_depth?: number;
  allowed_relationship_types?: string[];
}): Promise<GraphPathResponse> {
  return request<GraphPathResponse>('/knowledge-graph/path', {
    method: 'POST',
    body: JSON.stringify(payload),
  });
}

export async function analyzeRelationships(payload: {
  source_node_id: string;
  target_node_id: string;
  max_depth?: number;
}): Promise<RelationshipAnalysisResponse> {
  return request<RelationshipAnalysisResponse>('/knowledge-graph/analyze', {
    method: 'POST',
    body: JSON.stringify(payload),
  });
}

export async function getGraphContext(payload: {
  entity_names?: string[];
  node_ids?: string[];
  depth?: number;
  max_entities?: number;
}): Promise<GraphContextResponse> {
  return request<GraphContextResponse>('/knowledge-graph/context', {
    method: 'POST',
    body: JSON.stringify(payload),
  });
}

// ----------------------------------------------------------------------
// Document Graph API Calls
// ----------------------------------------------------------------------

export async function extractDocumentGraph(documentId: string): Promise<Record<string, any>> {
  return request<Record<string, any>>(`/knowledge-graph/documents/${documentId}/extract`, {
    method: 'POST',
  });
}

export async function getDocumentEntities(documentId: string): Promise<KnowledgeGraphNode[]> {
  return request<KnowledgeGraphNode[]>(`/knowledge-graph/documents/${documentId}/entities`, {
    method: 'GET',
  });
}

export async function getDocumentRelationships(documentId: string): Promise<KnowledgeGraphEdge[]> {
  return request<KnowledgeGraphEdge[]>(`/knowledge-graph/documents/${documentId}/relationships`, {
    method: 'GET',
  });
}

export async function rebuildDocumentGraph(documentId: string): Promise<Record<string, any>> {
  return request<Record<string, any>>(`/knowledge-graph/documents/${documentId}/rebuild`, {
    method: 'POST',
  });
}

// ----------------------------------------------------------------------
// Memory-Graph Synchronization API Calls
// ----------------------------------------------------------------------

export async function syncMemoryToGraph(memoryId: string): Promise<Record<string, any>> {
  return request<Record<string, any>>(`/knowledge-graph/sync/memory/${memoryId}`, {
    method: 'POST',
  });
}

export async function syncGraphNodeToMemory(nodeId: string): Promise<Record<string, any>> {
  return request<Record<string, any>>(`/knowledge-graph/nodes/${nodeId}/sync-memory`, {
    method: 'POST',
  });
}

